runtime/src/main/java/org/corfudb/runtime/object/MultiVersionObject.java

```diff
@@ -295,24 +301,25 @@ private void syncStreamUnsafe(long timestamp) {
                     try {
                         // Apply all updates in a MultiSMREntry, which is treated as one version.
                         final long globalAddress = addressUpdates.getGlobalAddress();
-                        addressUpdates.getSmrEntryList().forEach(this::applyUpdateUnsafe);
-
-                        final VersionedObjectIdentifier voId = new VersionedObjectIdentifier(getID(), globalAddress);
-
-                        // Populate the new version in the MVOCache and update version metadata.
-                        mvoCache.put(voId, currentObject);
-                        addressSpace.addAddress(globalAddress);
 
                         // The globalAddress can be equal to materializedUpTo when processing checkpoint
                         // entries that consist of multiple continuation entries. These will all share the
-                        // globalAddress of the no-op operation. There is no correctness issue by putting
-                        // these prematurely in the cache, as optimistic reads will be invalid.
+                        // globalAddress of the no-op operation. There is no correctness issue by prematurely
+                        // updating version information, as optimistic reads will be invalid.
                         Preconditions.checkState(globalAddress >= materializedUpTo,
                                 "globalAddress %s not >= materialized %s", globalAddress, materializedUpTo);
 
                         Preconditions.checkState(globalAddress >= resolvedUpTo,
                                 "globalAddress %s not >= resolved %s", globalAddress, resolvedUpTo);
 
+                        // If we observe a new version, place the previous one into the MVOCache.
+                        if (globalAddress > materializedUpTo) {
+                            final VersionedObjectIdentifier voId = new VersionedObjectIdentifier(getID(), materializedUpTo);
+                            mvoCache.put(voId, currentObject);
+                        }
+
+                        addressUpdates.getSmrEntryList().forEach(this::applyUpdateUnsafe);
```

Comment:
I think its noteworthy to mention that even though the hole doesn't mutate `currentObject` it can change its version. For example successive holes can generate multiple versions that point to the same exact object.

Reply:
Done.